config.org
#+title: Config

* Table of Content :toc:
- [[#configel][config.el]]
  - [[#name][Name]]
  - [[#ricing][Ricing]]
  - [[#org-mode][Org Mode]]
  - [[#latex][Latex]]
  - [[#corfu][Corfu]]

* config.el
:PROPERTIES:
:header-args: :tangle config.el
:END:
This section contains the configuration for ~doom emacs~ once the initial packages and modules have been loaded.

#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+end_src
** Name
#+begin_src elisp
(setq user-full-name "Bach W")
#+end_src
** Ricing
*** Theme
I like light theme
#+begin_src elisp
(setq doom-theme 'doom-flatwhite)
#+end_src
*** Font
Currently, we like to use IBM Plex Mono, but we would like this configuration to only load that theme if it is present on the system.

#+begin_src elisp
(when (doom-font-exists-p "BlexMono Nerd Font Mono")
  (setq doom-font (font-spec :family "BlexMono Nerd Font Mono" :size 14)))
(when (doom-font-exists-p "IBM Plex Sans")
  (setq doom-variable-pitch-font (font-spec :family "IBM Plex Sans" :size 14)))
#+end_src
*** Dashboard
#+begin_src elisp
(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)
(add-hook! '+doom-dashboard-functions :append (insert "\n" (+doom-dashboard--center +doom-dashboard--width "Best Editor in History!")))
;; Replacing logo with image
(setq fancy-splash-image (concat doom-private-dir "/images/emacs.svg"))
#+end_src

*** Line Numbers

#+begin_src elisp
(setq display-line-numbers-type 'relative)
#+end_src

** Org Mode
*** Basic
Firstly, we have to setup the org-directory to match with our system convention.
#+begin_src elisp
(setq org-directory "~/org/")
#+end_src

#+begin_src elisp
(setq-default org-download-image-dir "~/org/assets/images")
#+end_src
*** Roam
**** Additional Templates

#+caption: Description of the different templates.
| Name       | Hot Key | Description                                       |
|------------+---------+---------------------------------------------------|
| Encryption | e       | Used to generate encrypted ~org-roam~ entries.      |
|            |         | This is required for notes that cannot be seen by |
|            |         | anyone else.                                      |
|------------+---------+---------------------------------------------------|
#+begin_src elisp
(defun my/org-roam-load-templates ()
  (add-to-list 'org-roam-capture-templates
               '("e" "encrypted" plain "%?"
                 :target (file+head "%<%Y%m%d%H%M%S>-${slug}.org.gpg"
                                    "#+title: ${title}\n")
                 :unnarrowed t)))
#+end_src

**** Agenda Synchronization
The following configuration allows for notes in ~org-roam~ to be used as a task reference page.
#+begin_src elisp
(defun my/org-roam-create-filter-by-tag (tag-name)
  (lambda (node)
    (member tag-name (org-roam-node-tags node))))

(defun my/org-roam-list-notes-by-tag (tag-name)
  (mapcar #'org-roam-node-file
          (seq-filter (my/org-roam-create-filter-by-tag tag-name)
                      (org-roam-node-list))))

(defun my/org-agenda-refresh-list ()
  "function that reloads the notes registered to agenda. Looks for tracktodo"
  (interactive)
  (setq org-agenda-files (append
                          (my/org-roam-list-notes-by-tag "tracktodo")
                          '("~/org/todo.org"
                            "~/org/schedule.org"
                            "~/org/agenda/events.org"
                            "~/org/agenda/birthdays.org"
                            "~/org/agenda/reading-list.org"))))

(defun my/org-roam-copy-todo-to-today ()
  "function that registers the hook to move done task to today"
  (interactive)
  (let ((org-refile-keep t) ;; Set this to nil to delete the original!
        (org-roam-dailies-capture-templates
         '(("t" "tasks" entry "%?"
            :if-new (file+head+olp "%<%Y-%m-%d>.org" "#+title: %<%Y-%m-%d>\n" ("Tasks")))))
        (org-after-refile-insert-hook #'save-buffer)
        today-file
        pos)
    (save-window-excursion
      (org-roam-dailies--capture (current-time) t)
      (setq today-file (buffer-file-name))
      (setq pos (point)))

    ;; Only refile if the target file is different than the current file
    (unless (equal (file-truename today-file)
                   (file-truename (buffer-file-name)))
      (org-refile nil nil (list "Tasks" today-file nil pos)))))
#+end_src
**** Hook registration
In this section, the relevant hooks for ~org-roam~ are registers so that it would load after the package has been loaded by ~doom emacs~.
#+begin_src elisp
(after! org-roam
  (my/org-agenda-refresh-list)
  (add-to-list 'org-after-todo-state-change-hook
               (lambda ()
                 (when (equal org-state "DONE")
                   (my/org-roam-copy-todo-to-today))))
  (setq org-startup-folded 'content
        org-log-into-drawer nil)
  (my/org-roam-load-templates))
#+end_src

*** Agenda
#+begin_src elisp
(setq org-agenda-start-with-log-mode t)
(setq org-log-done 'time)
#+end_src
*** Pomodoro
Here, we could determine the timing of Pomodoro using our system.
#+begin_src elisp
(setq org-pomodoro-length 50)
(setq org-pomodoro-short-break-length 10)
#+end_src
** Latex
#+caption: This snippet is used to change the priority of tab completion when in LaTeX mode.
#+begin_src elisp
(map! :map cdlatex-mode-map
      :i "TAB" #'cdlatex-tab)
#+end_src
** Corfu
#+begin_src elisp
(after! corfu
  (setq +corfu-want-tab-prefer-expand-snippets t
        +corfu-want-tab-prefer-navigating-snippets t
        +corfu-want-tab-prefer-navigating-org-tables t))
#+end_src


config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

(setq user-full-name "Bach W")

(setq doom-theme 'doom-flatwhite)

(when (doom-font-exists-p "BlexMono Nerd Font Mono")
  (setq doom-font (font-spec :family "BlexMono Nerd Font Mono" :size 14)))
(when (doom-font-exists-p "IBM Plex Sans")
  (setq doom-variable-pitch-font (font-spec :family "IBM Plex Sans" :size 14)))

(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)
(add-hook! '+doom-dashboard-functions :append (insert "\n" (+doom-dashboard--center +doom-dashboard--width "Best Editor in History!")))
;; Replacing logo with image
(setq fancy-splash-image (concat doom-private-dir "/images/emacs.svg"))

(setq display-line-numbers-type 'relative)

(setq org-directory "~/org/")

(setq-default org-download-image-dir "~/org/assets/images")

(defun my/org-roam-load-templates ()
  (add-to-list 'org-roam-capture-templates
               '("e" "encrypted" plain "%?"
                 :target (file+head "%<%Y%m%d%H%M%S>-${slug}.org.gpg"
                                    "#+title: ${title}\n")
                 :unnarrowed t)))

(defun my/org-roam-create-filter-by-tag (tag-name)
  (lambda (node)
    (member tag-name (org-roam-node-tags node))))

(defun my/org-roam-list-notes-by-tag (tag-name)
  (mapcar #'org-roam-node-file
          (seq-filter (my/org-roam-create-filter-by-tag tag-name)
                      (org-roam-node-list))))

(defun my/org-agenda-refresh-list ()
  "function that reloads the notes registered to agenda. Looks for tracktodo"
  (interactive)
  (setq org-agenda-files (append
                          (my/org-roam-list-notes-by-tag "tracktodo")
                          '("~/org/todo.org"
                            "~/org/schedule.org"
                            "~/org/agenda/events.org"
                            "~/org/agenda/birthdays.org"
                            "~/org/agenda/reading-list.org"))))

(defun my/org-roam-copy-todo-to-today ()
  "function that registers the hook to move done task to today"
  (interactive)
  (let ((org-refile-keep t) ;; Set this to nil to delete the original!
        (org-roam-dailies-capture-templates
         '(("t" "tasks" entry "%?"
            :if-new (file+head+olp "%<%Y-%m-%d>.org" "#+title: %<%Y-%m-%d>\n" ("Tasks")))))
        (org-after-refile-insert-hook #'save-buffer)
        today-file
        pos)
    (save-window-excursion
      (org-roam-dailies--capture (current-time) t)
      (setq today-file (buffer-file-name))
      (setq pos (point)))

    ;; Only refile if the target file is different than the current file
    (unless (equal (file-truename today-file)
                   (file-truename (buffer-file-name)))
      (org-refile nil nil (list "Tasks" today-file nil pos)))))

(after! org-roam
  (my/org-agenda-refresh-list)
  (add-to-list 'org-after-todo-state-change-hook
               (lambda ()
                 (when (equal org-state "DONE")
                   (my/org-roam-copy-todo-to-today))))
  (setq org-startup-folded 'content
        org-log-into-drawer nil)
  (my/org-roam-load-templates))

(setq org-agenda-start-with-log-mode t)
(setq org-log-done 'time)

(setq org-pomodoro-length 50)
(setq org-pomodoro-short-break-length 10)

(map! :map cdlatex-mode-map
      :i "TAB" #'cdlatex-tab)

(after! corfu
  (setq +corfu-want-tab-prefer-expand-snippets t
        +corfu-want-tab-prefer-navigating-snippets t
        +corfu-want-tab-prefer-navigating-org-tables t))
